File: HW4/plot.R
# Jakob Balkovec
# CPSC 4330
#     Mar 4th 2025
#
# plot.R  [alias: hw4.R]
#   Data visualization for the K-Means algorithm
#
# To run:
#   >>> source("plot.R")

library(ggplot2)
library(dplyr)

# Assuming `device_status` is a dataframe with columns V4 (lat) and V5 (long)
coordinates <- device_status %>%
  select(lat = V4, long = V5) %>%
  filter(lat != 0 & long != 0)  # Filter out locations with lat/long equal to 0

# K-means parameters
K <- 5
convergeDist <- 0.1

# Randomly select K points as the initial centroids
set.seed(34)  # For reproducibility
initial_centroids <- coordinates[sample(1:nrow(coordinates), K), ]

# Function to calculate Euclidean distance
euclidean_distance <- function(p1, p2) {
  sqrt(sum((p1 - p2)^2))
}

# Function to assign clusters
assign_clusters <- function(data, centroids) {
  clusters <- vector("list", length = K)
  for (i in 1:nrow(data)) {
    distances <- apply(centroids, 1, function(centroid) euclidean_distance(data[i,], centroid))
    cluster <- which.min(distances)
    clusters[[cluster]] <- append(clusters[[cluster]], i)
  }
  return(clusters)
}

# Function to update centroids
update_centroids <- function(data, clusters) {
  new_centroids <- matrix(NA, nrow = K, ncol = 2)
  for (i in 1:K) {
    cluster_data <- data[clusters[[i]], ]
    new_centroids[i, ] <- colMeans(cluster_data)
  }
  return(new_centroids)
}

# K-means algorithm
centroids <- as.matrix(initial_centroids)
converged <- FALSE
while (!converged) {
  # Assign points to the nearest centroid
  clusters <- assign_clusters(coordinates, centroids)
  
  # Update centroids
  new_centroids <- update_centroids(coordinates, clusters)
  
  # Check for convergence
  dist_movement <- sum(apply(new_centroids - centroids, 1, function(x) sum(x^2)))
  if (dist_movement < convergeDist) {
    converged <- TRUE
  }
  
  # Update centroids for the next iteration
  centroids <- new_centroids
}

# Print the final centroids
print("Final Centroids:")
print(centroids)

# Plot the data and centroids
coordinates$cluster <- factor(unlist(lapply(1:K, function(i) rep(i, length(clusters[[i]])))))
ggplot(coordinates, aes(x = lat, y = long, color = cluster)) +
  geom_point(alpha = 0.6) +
  geom_point(data = as.data.frame(centroids), aes(x = V1, y = V2), color = "red", size = 5, shape = 4) +
  theme_minimal() +
  ggtitle("K-Means Clustering of Device Locations") +
  xlab("Latitude") +
  ylab("Longitude")

File: HW4/plot.py
# Jakob Balkovec
# CPSC 4330
#     Mar 4th 2025
#
# plot.py  [alias: hw4.py]
#   Data visualization for the K-Means algorithm
#
# To run:
#   >>> python plot.py
#   

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.cluster import KMeans

# Constants
FILE_NAME = r"/Users/jbalkovec/Desktop/CPSC4330/HW4/devicestatus-clean/devicestatus-mini.txt"
NUM_CLUSTERS = 5  # Adjust as needed

def parse_line(line):
    """
    Parses a line from the dataset and extracts latitude and longitude.
    
    Parameters:
        line (str): A single line from the dataset.
    
    Returns:
        tuple: (latitude, longitude) if valid, otherwise None.
    """
    try:
        parts = line.strip().split(",")
        if len(parts) < 5:
            return None

        lat, lon = float(parts[3]), float(parts[4])

        # Ignore invalid coordinates (0.0, 0.0)
        if lat == 0.0 and lon == 0.0:
            return None
        
        return (lat, lon)
    
    except ValueError as e:
        print(f"Error parsing line: {line.strip()} - {e}")
        return None

def load_coordinates(file_path):
    """
    Reads and parses the dataset to extract valid latitude-longitude pairs.
    
    Parameters:
        file_path (str): Path to the dataset file.
    
    Returns:
        pd.DataFrame: DataFrame containing latitude and longitude columns.
    """
    with open(file_path, 'r') as f:
        lines = f.readlines()
        coordinates = [parse_line(x) for x in lines if parse_line(x) is not None]

    return pd.DataFrame(coordinates, columns=["Latitude", "Longitude"])

def perform_kmeans(data, num_clusters=NUM_CLUSTERS):
    """
    Performs K-means clustering on the given dataset.
    
    Parameters:
        data (pd.DataFrame): DataFrame containing 'Latitude' and 'Longitude'.
        num_clusters (int): Number of clusters for K-means.
    
    Returns:
        tuple: (Updated DataFrame with cluster labels, cluster centroids)
    """
    kmeans = KMeans(n_clusters=num_clusters, random_state=42, n_init=10)
    data["Cluster"] = kmeans.fit_predict(data[["Latitude", "Longitude"]])
    return data, kmeans.cluster_centers_

def plot_clusters(data, centroids, num_clusters=NUM_CLUSTERS):
    """
    Plots clustered points along with centroids.
    
    Parameters:
        data (pd.DataFrame): DataFrame containing clustered location data.
        centroids (ndarray): Coordinates of cluster centroids.
        num_clusters (int): Number of clusters.
    """
    plt.figure(figsize=(10, 7))
    sns.scatterplot(x="Latitude", y="Longitude", hue="Cluster", palette="tab10", data=data, alpha=0.7, edgecolor="k")
    plt.scatter(centroids[:, 0], centroids[:, 1], color="red", marker="X", s=200, label="Centroids")
    
    plt.xlabel("Latitude", fontsize=12)
    plt.ylabel("Longitude", fontsize=12)
    plt.title(f"K-Means Clustering of Device Locations (K={num_clusters})", fontsize=14, fontweight="bold")
    plt.legend()
    plt.grid(True, linestyle="--", alpha=0.6)
    plt.show()

# Main Execution
if __name__ == "__main__":
    coordinates_df = load_coordinates(FILE_NAME)
    clustered_data, centroids = perform_kmeans(coordinates_df)
    plot_clusters(clustered_data, centroids)